_posts/2023-03-13-llvm-opaque-pointers.md

```diff
@@ -0,0 +1,85 @@
+---
+title: "LLVM opaque pointer support has landed"
+author: HertzDevil
+summary: 
+---
+
+Crystal 1.8, the upcoming minor release, will support LLVM's opaque pointers for the first time, allowing the compiler to be built with LLVM 15 or above. To understand the significance of opaque pointers, let's take a look at a small sample program:
+
+```crystal
+# test.cr
+class Foo
+  def initialize(@x : Int32)
+  end
+end
+
+Foo.new(1)
+```
+
+Build the above program with `crystal build --prelude=empty --no-debug --emit=llvm-ir test.cr`. The compiler will create a file `test.ll` containing our program's LLVM IR, the platform-independent intermediate representation used by LLVM to emit LLVM bytecode and eventually machine code. The following is the LLVM function corresponding to `Foo#initialize`:
+
+```llvm
+; Function Attrs: uwtable
+define internal i32 @"*Foo#initialize<Int32>:Int32"(%Foo* %self, i32 %x) #0 {
+entry:
+  %0 = getelementptr inbounds %Foo, %Foo* %self, i32 0, i32 1
+  store i32 %x, i32* %0, align 4
+  ret i32 %x
+}
+```
+
+Without going into the details of how exactly Crystal compiles the method into this LLVM function (although we did actually have a [write-up](https://crystal-lang.org/2015/03/04/internals/) on this in the past), we know that it:
+
+* Takes as arguments a `%self` parameter, the `Foo` object being constructed, and the `%x` parameter coming from `Foo#initialize`;
+* Assigns the address of the `Foo` object's `@x` instance variable into the local variable `%0`;
+* Stores the `%x` parameter into `@x` via `%0`;
+* Returns `%x` to the caller. (This caller is normally `Foo.new`, so it goes unused most of the time.)
+
+We can see that the types of `%self` and `%0` are `%Foo*` and `i32*` respectively. These are _typed pointers_, which LLVM has been using for a long time. In LLVM, typed pointers of different pointee types must be explicitly converted using the `bitcast` LLVM instruction, otherwise the resulting LLVM IR is ill-formed. As more and more LLVM frontends came into existence, it was soon realized that typed pointers do not offer many useful semantics, and instead add an unnecessary layer of complexity over IR generation and analysis.
+
+The _opaque pointer_ was first suggested back in [February 2015](https://lists.llvm.org/pipermail/llvm-dev/2015-February/081822.html), where all pointer types would be represented by a single `ptr` in LLVM IR. Then in September 2022, [LLVM 15 now uses opaque pointers by default](https://releases.llvm.org/15.0.0/docs/OpaquePointers.html#version-support), and typed pointers will be removed shortly afterwards. If we compile the program above again, but this time with a Crystal compiler built using LLVM 15, we can see the opaque pointers in action:
```

Comment:
Maybe this could be a good place to insert a section headline?

```suggestion
## Opaque Pointers

The _opaque pointer_ was first suggested back in [February 2015](https://lists.llvm.org/pipermail/llvm-dev/2015-February/081822.html), where all pointer types would be represented by a single `ptr` in LLVM IR. Then in September 2022, [LLVM 15 now uses opaque pointers by default](https://releases.llvm.org/15.0.0/docs/OpaquePointers.html#version-support), and typed pointers will be removed shortly afterwards. If we compile the program above again, but this time with a Crystal compiler built using LLVM 15, we can see the opaque pointers in action:
```